bool rightSide
float32 angle_new
int32 sign